pkg/devfile/parser/parse.go

```diff
@@ -242,53 +281,134 @@ func parseFromURI(uri string, curDevfileCtx devfileCtx.DevfileCtx) (DevfileObj,
+	id := importReference.Id
+	registryURL := importReference.RegistryUrl
 	if registryURL != "" {
-		devfileContent, err := getDevfileFromRegistry(parentId, registryURL)
+		devfileContent, err := getDevfileFromRegistry(id, registryURL)
 		if err != nil {
 			return DevfileObj{}, err
 		}
-		return ParseDevfile(ParserArgs{Data: devfileContent, RegistryURLs: curDevfileCtx.GetRegistryURLs()})
-	} else if curDevfileCtx.GetRegistryURLs() != nil {
-		for _, registry := range curDevfileCtx.GetRegistryURLs() {
-			devfileContent, err := getDevfileFromRegistry(parentId, registry)
+		err = d.Ctx.SetDevfileContentFromBytes(devfileContent)
+		if err != nil {
+			return d, errors.Wrap(err, "failed to set devfile content from bytes")
+		}
+		newResolveCtx := resolveCtx.appendNode(importReference)
+
+		return populateAndParseDevfile(d, newResolveCtx, tool, true)
+
+	} else if tool.registryURLs != nil {
+		for _, registryURL := range tool.registryURLs {
+			devfileContent, err := getDevfileFromRegistry(id, registryURL)
 			if devfileContent != nil && err == nil {
-				return ParseDevfile(ParserArgs{Data: devfileContent, RegistryURLs: curDevfileCtx.GetRegistryURLs()})
+				err = d.Ctx.SetDevfileContentFromBytes(devfileContent)
+				if err != nil {
+					return d, errors.Wrap(err, "failed to set devfile content from bytes")
+				}
+				importReference.RegistryUrl = registryURL
+				newResolveCtx := resolveCtx.appendNode(importReference)
+
+				return populateAndParseDevfile(d, newResolveCtx, tool, true)
 			}
 		}
 	} else {
 		return DevfileObj{}, fmt.Errorf("failed to fetch from registry, registry URL is not provided")
 	}
 
-	return DevfileObj{}, fmt.Errorf("failed to get parent Id: %s from registry URLs provided", parentId)
+	return DevfileObj{}, fmt.Errorf("failed to get id: %s from registry URLs provided", id)
 }
 
-func getDevfileFromRegistry(parentId, registryURL string) ([]byte, error) {
+func getDevfileFromRegistry(id, registryURL string) ([]byte, error) {
 	if !strings.HasPrefix(registryURL, "http://") && !strings.HasPrefix(registryURL, "https://") {
 		return nil, fmt.Errorf("the provided registryURL: %s is not a valid URL", registryURL)
 	}
 	param := util.HTTPRequestParams{
-		URL: fmt.Sprintf("%s/devfiles/%s", registryURL, parentId),
+		URL: fmt.Sprintf("%s/devfiles/%s", registryURL, id),
 	}
 	return util.HTTPGetRequest(param, 0)
 }
+
+func parseFromKubeCRD(importReference v1.ImportReference, resolveCtx *resolutionContextTree, tool resolverTools) (d DevfileObj, err error) {
+
+	if tool.k8sClient == nil || tool.context == nil {
+		return DevfileObj{}, fmt.Errorf("Kubernetes client and context are required to parse from Kubernetes CRD")
+	}
+	namespace := importReference.Kubernetes.Namespace
+	// if namespace is not set in devfile, use default namespace provided in by consumer
+	if namespace == "" {
+		namespace = tool.defaultNamespace
+	}
+	// use current namespace if namespace is not set in devfile and not provided by consumer
+	if namespace == "" {
```

Comment:
can we update this? we have two `if namespace == "" {` one after the other

Reply:
updated